### PR TITLE
Allow different shell runners

### PR DESCRIPTION
**Pull request by vknabel**
### About

This PR allows the SwiftShell implementation to be replaced by a custom one.
Different tools (example below) and unit tests (asserting what to be run) could benefit from this.

Not sure if you want this or if they fit into this project's conventions.

### Use case
For Archery I plan to run swift tools using Mint in order to implement sources/loaders for the Archerfile. Other swift tools (arrows) would then be run on the combined data. (e.g. one could add more scripts from a source like Beak, cmdshelf or to share something like common scripts)
See [vknabel/Archery#archerfile-loaders](https://github.com/vknabel/Archery#archerfile-loaders)

### CODE REVIEW

File: Sources/MintKit/CustomShell.swift [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import SwiftShell
+
+public protocol CustomShell {
+    func run(bash: String, env: [String: String], directory: String?) -> RunOutput
+    func runAndPrint(bash: String, env: [String: String], directory: String?) throws
```
Comment (toshi0383) [inline Sources/MintKit/CustomShell.swift]: I don't think this one is necessary. It's same as the one below, isn't it?

File: Sources/MintKit/Mint.swift [hunk 2]
```diff
@@ -99,12 +100,11 @@ public struct Mint {
...
 
-        var context = CustomContext(main)
-        context.env["MINT"] = "YES"
-        context.env["RESOURCE_PATH"] = ""
-
         let packagePath = PackagePath(path: packagesPath, package: package)
-        try context.runAndPrint(packagePath.commandPath.string, arguments)
+        try shell.runAndPrint(packagePath.commandPath.string, arguments, env: [
+            "MINT": "YES",
+            "RESOURCE_PATH": ""
```
Comment (toshi0383) [inline Sources/MintKit/Mint.swift]: I guess we can move these env variables into the default implementation, but in different PR.

File: Sources/MintKit/CustomShell.swift (hunk 1 above)
Comment (vknabel) [reply Sources/MintKit/CustomShell.swift]: Thank you for your review.

Regarding `runAndPrint` there is a slight difference between the variants.
`runAndPrint(bash: "echo hello world")` will be executed as `/bin/bash -c "echo hello world"`,
whereas `runAndPrint(_:_:env:directory:)` requires a full path like `runAndPrint("/bin/bash", ["-c", "echo hello world"])`. Within Mint itself, it is used to run the installed packed.

Maybe we should rename one variant?

**Comment by yonaskolb** [comment]
Thanks for the changes @vknabel.
I'm going to look at replacing `SwiftShell` with `SwiftCLI`, and will then revisit this

**Comment by vknabel** [comment]
@yonaskolb just curious: why do you think of replacing `SwiftShell` with `SwiftCLI`?

**Comment by yonaskolb** [comment]
The process executing support is a little more full featured, including support for interrupts. It also means it can replace SPM ArgumentParser, as it has a really nice API for commands and arguments.

It's not that all that different in the end, but just want to give it a go. Might not integrate it after trying  it out.

**Comment by vknabel** [comment]
I guess this can be closed. It probably isn't relevant anymore.
Keep up your good work!

**Pull request closed without merging**